qiskit/optimization/algorithms/optimization_algorithm.py

```diff
@@ -101,112 +103,121 @@ class OptimizationResult:
     Attributes:
         x: The optimal value found in the optimization algorithm.
         fval: The function value corresponding to the optimal value.
-        results: The original results object returned from the optimization algorithm. This can
+        raw_results: The original results object returned from the optimization algorithm. This can
             contain more information than only the optimal value and function value.
         status: The termination status of the algorithm.
         variables: The list of variables under optimization.
+        variable_names: The list of variable names.
+        variables_dict: The optimal value as a dictionary of the variable name and corresponding
+            value.
+
+    `OptimizationResult` allows users to get the value of a variable by specifying an index or
+    a name as follows.
+
+    Examples:
+        >>> problem = QuadraticProgram()
+        >>> problem.binary_var('x1')
+        >>> problem.binary_var('x2')
+        >>> problem.binary_var('x3')
+        >>> problem.minimize(linear={'x1': 1, 'x2': -2, 'x3': 3})
+        >>> print([var.name for var in problem.variables])
+        ['x1', 'x2', 'x3']
+        >>> algo = CplexOptimizer()
+        >>> result = algo.solve(problem)
+        >>> print(result.variable_names)
+        ['x1', 'x2', 'x3']
+        >>> print(result.x)
+        [0.0, 1.0, 0.0]
+        >>> print(result[1])
+        1.0
+        >>> print(result['x1'])
+        0.0
+        >>> print(result.fval)
+        -2.0
+        >>> print(result.variables_dict)
+        {'x1': 0.0, 'x2': 1.0, 'x3': 0.0}
+
+    Note:
+        The order of variables are supposed to be equal to that of the problem solved by
+        optimization algorithms. Optimization algorithms and converters of `QuadraticProgram`
+        should maintain the order when generating a new `OptimizationResult` object.
     """
 
     Status = OptimizationResultStatus
 
-    def __init__(self, x: Optional[Any] = None, fval: Optional[Any] = None,
-                 results: Optional[Any] = None,
-                 status: OptimizationResultStatus = OptimizationResultStatus.SUCCESS,
-                 variables: Optional[List[Variable]] = None) -> None:
-        self._x = x if x is not None else []   # pylint: disable=invalid-name
-        self._variables = variables if variables is not None else []
+    def __init__(self, x: Union[List[float], np.ndarray], fval: float,
+                 variables: List[Variable],
+                 raw_results: Optional[Any] = None,
```

Comment:
Should there be some structure to raw_results, as in rather than `Any` would `Dict[str, Any]` be preferable? I see this for instance the GroverOptimizer passes in a dictionary there. Not sure for other cases. At least the string field names give some clue to the content which might not otherwise be there if its Any directly.

Reply:
I'd suggest to keep it as is now and return to `raw_result` in #1120, if you don't mind.

Reply:
I'd like to keep `raw_results` just `Any` in this PR and discuss details of raw_results in #1120 too. Is it OK?

Reply:
Sure, was hoping that the Dict would cover things, but we can leave it as-is and visit it as part of that issue,